test/e2e/gardener/shoot/create_rotate_delete.go

```diff
@@ -105,62 +165,13 @@ var _ = Describe("Shoot Tests", Label("Shoot", "default"), func() {
 				v = append(v, &rotation.SSHKeypairVerifier{ShootCreationFramework: f})
 			}
 
-			DeferCleanup(func() {
-				ctx, cancel := context.WithTimeout(parentCtx, 2*time.Minute)
-				defer cancel()
-
-				v.Cleanup(ctx)
-			})
-
-			v.Before(ctx)
+			// test rotation for every rotation type
+			testCredentialRotation(ctx, v, f)
 
-			By("Start credentials rotation")
-			ctx, cancel = context.WithTimeout(parentCtx, 20*time.Minute)
-			defer cancel()
-
-			patch := client.MergeFrom(f.Shoot.DeepCopy())
-			metav1.SetMetaDataAnnotation(&f.Shoot.ObjectMeta, v1beta1constants.GardenerOperation, v1beta1constants.OperationRotateCredentialsStart)
-			Eventually(func() error {
-				return f.GardenClient.Client().Patch(ctx, f.Shoot, patch)
-			}).Should(Succeed())
-
-			Eventually(func(g Gomega) {
-				g.Expect(f.GardenClient.Client().Get(ctx, client.ObjectKeyFromObject(f.Shoot), f.Shoot)).To(Succeed())
-				g.Expect(f.Shoot.Annotations).NotTo(HaveKey(v1beta1constants.GardenerOperation))
-				v.ExpectPreparingStatus(g)
-			}).Should(Succeed())
-
-			Expect(f.WaitForShootToBeReconciled(ctx, f.Shoot)).To(Succeed())
-
-			Eventually(func() error {
-				return f.GardenClient.Client().Get(ctx, client.ObjectKeyFromObject(f.Shoot), f.Shoot)
-			}).Should(Succeed())
-
-			v.AfterPrepared(ctx)
-
-			By("Complete credentials rotation")
-			ctx, cancel = context.WithTimeout(parentCtx, 20*time.Minute)
-			defer cancel()
-
-			patch = client.MergeFrom(f.Shoot.DeepCopy())
-			metav1.SetMetaDataAnnotation(&f.Shoot.ObjectMeta, v1beta1constants.GardenerOperation, v1beta1constants.OperationRotateCredentialsComplete)
-			Eventually(func() error {
-				return f.GardenClient.Client().Patch(ctx, f.Shoot, patch)
-			}).Should(Succeed())
-
-			Eventually(func(g Gomega) {
-				g.Expect(f.GardenClient.Client().Get(ctx, client.ObjectKeyFromObject(f.Shoot), f.Shoot)).To(Succeed())
-				g.Expect(f.Shoot.Annotations).NotTo(HaveKey(v1beta1constants.GardenerOperation))
-				v.ExpectCompletingStatus(g)
-			}).Should(Succeed())
-
-			Expect(f.WaitForShootToBeReconciled(ctx, f.Shoot)).To(Succeed())
-
-			Eventually(func(g Gomega) {
-				g.Expect(f.GardenClient.Client().Get(ctx, client.ObjectKeyFromObject(f.Shoot), f.Shoot)).To(Succeed())
-			}).Should(Succeed())
-
-			v.AfterCompleted(ctx)
+			// isolated test for ssh key rotation (does not trigger node rolling update)
```

Comment:
I'm not sure whether you are testing the intended thing here. 
For ssh key rotation only, you have to annotate the shoot with [`gardener.cloud/operation=rotate-ssh-keypair`](https://github.com/gardener/gardener/blob/master/docs/usage/shoot-operations/shoot_credentials_rotation.md#ssh-key-pair-for-worker-nodes) while `testCredentialsRotation` function is triggering a complete credentials rotation.

I'm not entirely sure about the reason for the failing tests, but the second rotation might be related. During credentials rotation kube-apiserver is restarting a couple of times. When you do it two times in a row, it can end in a crashloop backoff.

Reply:
Oh, yeah!
I completely oversaw the triggering of the credential rotation [here](https://github.com/gardener/gardener/pull/10671/files#diff-8754abb9e4269533fe6255890ad26eff40d799ea5fe08b9f40329394052541f1R38) using the `v1beta1constants.OperationRotateCredentialsStart`...

I will refactor this and try to reproduce locally.
Thank you very much for the insight🚀

Reply:
I refactored it locally, but the e2e test still fails.

Contrary to expectation, it fails during the full credential rotation, not in the second one.